### PR TITLE
Make use of `_parameter_upper_bound` safer

### LINKED ISSUE #208: `Base.parameter_upper_bound` was removed in Julia v1.6-DEV
`Base.parameter_upper_bound` was removed in https://github.com/JuliaLang/julia/pull/32427. It is used in:
https://github.com/JuliaGraphics/ColorTypes.jl/blob/368bcd0408e0ff8701e91d80e1deeaecb00083d0/src/traits.jl#L132-L134
https://github.com/JuliaGraphics/ColorTypes.jl/blob/368bcd0408e0ff8701e91d80e1deeaecb00083d0/src/traits.jl#L168-L171
https://github.com/JuliaGraphics/ColorTypes.jl/blob/368bcd0408e0ff8701e91d80e1deeaecb00083d0/src/traits.jl#L230

To be fair, it had been a non-public API since Julia v0.7. However, `Base.unwrap_unionall` is also not a public API.

### PR DESCRIPTION
**Pull request by kimikage**
This makes sure that `_parameter upper_bound` is used with `TransparentColor`, not its subtypes.
This also changes the implementation of `eltypes_supported`.

Closes #208

### CODE REVIEW

File: src/traits.jl [hunk 1]
```diff
@@ -129,15 +129,16 @@ eltype(::Type{Colorant{T,N}}) where {T,N}   = T
...
 # eltypes_supported(Colorant{T<:X}) -> X
 function eltypes_supported(::Type{C}) where {C<:Colorant}
     Cb = base_colorant_type(C)
     isconcretetype(C) && C === Cb && return eltype(C)
-    _parameter_upper_bound(Cb, 1)
+    _eltypes_supported(Cb)
+end
+@pure function _eltypes_supported(::Type{C}) where {C<:Colorant}
+    nameof(C) === :Colorant || return _eltypes_supported(supertype(C))
+    C.var.name === :T || return Any
```
Comment (timholy) [inline src/traits.jl:140]: Depending on `C.var.name` seems a little risky, though surprisingly it works for now:
```julia
julia> abstract type Container{T} end

julia> foo(::Type{C}) where C<:Container = C.var.name
foo (generic function with 1 method)

julia> foo(Container)
:T

julia> foo(Container{U} where U)
:T

julia> C = Container{U} where U
Container{U} where U

julia> C.var.name
:U
```

File: test/traits.jl [hunk 2]
```diff
@@ -224,11 +233,25 @@ end
...
+    @test @inferred(!ColorTypes.issupported(AGray, Bool))
+
+    @test @inferred(ColorTypes.eltypes_supported(C2{Float32})) === Real
+    @test_broken @inferred(ColorTypes.eltypes_supported(C2A)) === Real
+
+    @test @inferred(ColorTypes.eltypes_supported(TransparentColor)) === Any
+
+    ST = @inferred(ColorTypes.eltypes_supported(StrangeGray{:X}))
+    @test_broken ST === (Normed{T} where T <: Integer)
+    @test_broken eltype(StrangeGray{UInt16}(1N14f2)) isa ST
```
Comment (timholy) [inline test/traits.jl:251]:
```suggestion
    @test_broken eltype(StrangeGray{UInt16}(1N14f2)) <: ST
```
but I think this will pass?

Comment (kimikage) [reply test/traits.jl:251]: I intended to change the `eltype` to `gray`. It seems I forgot that while doing other work. :sweat_smile:
In any case, this is a `broken` for now.

File: src/traits.jl (hunk 1 above)
Comment (kimikage) [reply src/traits.jl:140]: Yes, this is a makeshift solution. I'm looking for a better way this past week.

Comment (kimikage) [reply src/traits.jl]: I changed the code to use `_parameter_upper_bound` again. However, the new implementation makes sure the input is a `Colorant`, so it is safer and more reasonable than the current master.

```julia
julia> ColorTypes.eltypes_supported(TransparentColor)
Color # current master
Any   # this PR

julia> ColorTypes.eltypes_supported(TransparentColor{RGB{T},T,4} where T) # analysis of constraints is still loose.
Any
```

**Comment by kimikage** [comment]
Thank you for the review. The last `push -f` is due to trivial code formatting.

I will merge this and #210, and then release v0.10.8. I believe it will be the last release of the v0.10 series, for sure. :sweat_smile:

**Pull request merged**